new-log-viewer/src/contexts/UrlContextProvider.tsx

```diff
@@ -0,0 +1,241 @@
+ * If a key's value is `null`, the key will be removed from the search parameters.
+ */
+const updateWindowSearchParams = (updates: UrlSearchParamUpdatesType) => {
+    const newUrl = new URL(window.location.href);
+    newUrl.search = getUpdatedSearchParams(updates);
+    window.history.pushState({}, "", newUrl);
+};
+
+/**
+ * Updates hash parameters in the current window's URL based on the provided key-value pairs.
+ *
+ * @param updates An object containing key-value pairs to update the hash parameters.
+ * If a key's value is `null`, the key will be removed from the hash parameters.
+ */
+const updateWindowHashParams = (updates: UrlHashParamUpdatesType) => {
+    const newHash = getUpdatedHashParams(updates);
+    const currHash = window.location.hash.substring(1);
+    if (newHash !== currHash) {
+        const newUrl = new URL(window.location.href);
+        newUrl.hash = newHash;
+        window.history.pushState({}, "", newUrl);
+        window.dispatchEvent(new HashChangeEvent("hashchange"));
+    }
+};
+
+/**
+ * Retrieves all search parameters from the current window's URL.
+ *
+ * @return An object containing the search parameters.
+ */
+const getAllWindowSearchParams = () => {
+    const urlSearchParams: NullableProperties<UrlSearchParams> = structuredClone(
+        URL_SEARCH_PARAMS_DEFAULT
+    );
+    const searchParams = new URLSearchParams(window.location.search.substring(1));
+
+    const filePath = searchParams.get(SEARCH_PARAM_NAME.FILE_PATH);
+    if (null !== filePath) {
+        urlSearchParams[SEARCH_PARAM_NAME.FILE_PATH] = getAbsoluteUrl(filePath);
+    }
+
+    return urlSearchParams;
+};
+
+/**
+ * Retrieves all hash parameters from the current window's URL.
+ *
+ * @return An object containing the hash parameters.
+ */
+const getAllWindowHashParams = () => {
+    const hashParams = new URLSearchParams(window.location.hash.substring(1));
+    const urlHashParams: NullableProperties<UrlHashParams> = structuredClone(
+        URL_HASH_PARAMS_DEFAULT
+    );
+
+    const logEventNum = hashParams.get(HASH_PARAM_NAME.LOG_EVENT_NUM);
+    if (null !== logEventNum) {
+        const parsed = Number(logEventNum);
+        urlHashParams[HASH_PARAM_NAME.LOG_EVENT_NUM] = isNaN(parsed) ?
+            null :
+            parsed;
+    }
+
+    return urlHashParams;
+};
+
+
+/**
+ * Copies the current window's URL to the clipboard. If any `updates` parameters are specified,
+ * the copied URL will include these modifications, while the original window's URL remains
+ * unchanged.
+ *
+ * @param searchParamUpdates An object containing key-value pairs to update the search parameters.
+ * If a key's value is `null`, the key will be removed from the search parameters.
+ * @param hashParamsUpdates An object containing key-value pairs to update the hash parameters.
+ * If a key's value is `null`, the key will be removed from the hash parameters.
+ */
+const copyToClipboard = (
```

Comment:
* Can we move this above the two methods above?
* How about `copyWindowUrlToClipboard`?

Reply:
I don't think `copyWindowUrlToClipboard` is a precise description of that function. How about `copyModifiedUrlToClipboard`?

Reply:
> I don't think copyWindowUrlToClipboard is a precise description of that function.

True.

> How about copyModifiedUrlToClipboard?

It's a little better, but I still don't think it necessarily gets the point across. How about `copyPermalinkToClipboard`?